### PR TITLE
refactor: use virtual environment to be consistent

### PR DESCRIPTION
**Pull request by SMoraisAnsys**
Following the work on #543, it seems that there was a need for three things:
- better handling of unexpected behavior when working with VMs, see https://github.com/ansys/actions/issues/559;
- better coherence in how we deal with `poetry` and `pip` build backends;
- better handling of the corner case where poetry and the project have shared dependencies with different versions.

This PR aims at tackling those three points by enforcing the use of a virtual environment, even when working with `poetry`.

On top of those changes, the modified actions are now "more compatible" with `poetry`. Before, the default behavior was to use `python -m pip install .` while now we detect the build backend and use `poetry` if appropriate.

> **Note**: following @Andy-Grigg comments, we use a "hack" to ensure that poetry and the project are not installed in the same environment and thus don't clash with each other if they share common dependencies with different version. The idea is to use `pipx` to install poetry in a temp folder cleaned up before and after running each job. The only side effect is for VMs as `pipx` will still be here afterward.

### CODE REVIEW

File: _doc-build-linux/action.yml [hunk 1]
```diff
@@ -179,56 +179,111 @@ runs:
...
-    - name: "Check if requirements.txt file exists"
+    # ------------------------------------------------------------------------
+
+    - uses: ansys/actions/_logging@main
+      with:
+        level: "INFO"
+        message: >
+          Determine context.
+
+    - name: "Determine Github environnement variables"
```
Comment (MaxJPRey) [inline _doc-build-linux/action.yml:190]:
```suggestion
    - name: "Determine GitHub environnement variables"
```

File: _doc-build-windows/action.yml [hunk 2]
```diff
@@ -211,47 +211,95 @@ runs:
...
+
+    - name: "Determine Github environnement variables"
+      shell: powershell
+      run: |
+        if ((Test-Path "pyproject.toml") -and (Get-Content "pyproject.toml" | Select-String -Pattern 'build-backend = "poetry\.core\.masonry\.api"')) {
+          echo "BUILD_BACKEND=$(echo 'poetry')" | Out-File -FilePath $env:GITHUB_ENV -Encoding utf8 -Append
+          echo "SPHINX_BUILD_MAKE=$(echo 'poetry run -- doc\make.bat')" | Out-File -FilePath $env:GITHUB_ENV -Encoding utf8 -Append
+        } else {
+          echo "BUILD_BACKEND=$(echo 'pip')" | Out-File -FilePath $env:GITHUB_ENV -Encoding utf8 -Append
+          echo "SPHINX_BUILD_MAKE=$(echo 'doc\make.bat')" | Out-File -FilePath $env:GITHUB_ENV -Encoding utf8 -Append
```
Comment (RobPasMue) [inline _doc-build-windows/action.yml:228]: Why the change of location for Windows? On Linux you were assuming the make commands to be ran at "doc" level. I would expect the same behavior on both actions.

**Review by RobPasMue — APPROVED** [review]
Overall LGTM - just a minor comment

File: _doc-build-windows/action.yml [hunk 3]
```diff
@@ -211,47 +211,95 @@ runs:
 
     # ------------------------------------------------------------------------
 
+    - uses: ansys/actions/_logging@main
+      with:
+        level: "INFO"
+        message: >
+          Determine context.
+
+    - name: "Determine Github environnement variables"
```
Comment (SMoraisAnsys) [inline _doc-build-windows/action.yml:220]:
```suggestion
    - name: "Determine GitHub environnement variables"
```

File: _doc-build-windows/action.yml [hunk 4]
```diff
@@ -211,47 +211,95 @@ runs:
 
     # ------------------------------------------------------------------------
 
+    - uses: ansys/actions/_logging@main
+      with:
+        level: "INFO"
+        message: >
+          Determine context.
+
+    - name: "Determine GitHub environnement variables"
```
Comment (SMoraisAnsys) [inline _doc-build-windows/action.yml:220]:
```suggestion
    - name: "Determine GitHub environment variables"
```

File: _doc-build-windows/action.yml (hunk 2 above)
Comment (SMoraisAnsys) [reply _doc-build-windows/action.yml:228]: The thing is that on Linux we leverage `make` with option `-C` to "change directory before doing anything". We can't do the same with Windows as we are directly calling the `make.bat` file.
If you want to have the same behavior for both actions, I can add a "move" step before building the documentation. Would you prefer that ?

Comment (RobPasMue) [reply _doc-build-windows/action.yml:228]: In my experience... context matters when running these batch script files. Sometimes they are not properly adapted to be ran from a different directory (which is what you are doing on the Windows version). I'd recommend changing directories, yeah.

Comment (SMoraisAnsys) [reply _doc-build-windows/action.yml:228]: Fine by me, in that case I'll avoid using `-C` with linux and also move to the `doc` folder. If you have any other feedback, let me now !

Comment (RobPasMue) [reply _doc-build-windows/action.yml:228]: ``make -C`` is effectively changing directory for you. so that's fine. You do not need to do it yourself on Linux. See https://www.man7.org/linux/man-pages/man1/make.1.html

make.bat is a batch script with which you are trying to mimic the Make GNU program. So, that's why the context is important in that case

Comment (SMoraisAnsys) [reply _doc-build-windows/action.yml:228]: While starting the modifications, I noticed that the steps associated to assets would be impacted by the changes.
Currently, this steps define `EXPECTED_BUILD_DIR=doc/_build` for Linux and `EXPECTED_BUILD_DIR=doc\_build` for Windows. While I do agree with your previous comment, I'd rather perform a manual folder move with `cd doc` in both Windows and Linux. Otherwise we would end up with "different" environment variable which might become confusing later on. @RobPasMue Is that fine with you ?

Comment (RobPasMue) [reply _doc-build-windows/action.yml:228]: Sure, go ahead

File: doc-build/action.yml [hunk 5]
```diff
@@ -220,7 +225,7 @@ runs:
 
     - name: Documentation build (Windows)
       if: ${{ runner.os == 'Windows' }}
-      uses: ansys/actions/_doc-build-windows@main
+      uses: ansys/actions/_doc-build-windows@fix/inconsistent-behavior
```
Comment (SMoraisAnsys) [inline doc-build/action.yml:228]:
```suggestion
      uses: ansys/actions/_doc-build-windows@main
```

File: tests-pytest/action.yml [hunk 6]
```diff
@@ -101,31 +101,72 @@ runs:
...
+        echo "POETRY_VIRTUALENVS_CREATE=false" >> $GITHUB_ENV
+
     - name: "Set up Python"
       uses: ansys/actions/_setup-python@main
       with:
         python-version: ${{ inputs.python-version }}
         use-cache: ${{ inputs.use-python-cache }}
 
-    - name: "Upgrade pip and disable virtual environment (for poetry only)"
+    - name: Create virtual environment
```
Comment (SMoraisAnsys) [inline tests-pytest/action.yml:153]:
```suggestion
    - name: "Create virtual environment"
```

File: _doc-build-windows/action.yml (hunk 2 above)
Comment (SMoraisAnsys) [reply _doc-build-windows/action.yml]: After testing I'll keep `make -C` for Linux (previous behavior). I though that it might be confusing with `EXPECTED_BUILD_DIR` but I forgot the fact that every step of the job starts at the root level. Even if in a previous step I did `cd doc` :)

**Comment by SMoraisAnsys** [comment]
This worked as expected for:
- https://github.com/ansys/pyansys-sound/actions/runs/10937085524
- https://github.com/ansys/pyconceptev/actions/runs/10936996838?pr=122
- currently running for https://github.com/ansys-internal/pyaedt-examples/actions/runs/10917092741?pr=230

**Comment by RobPasMue** [comment]
You sure had some fun in this PR @SMoraisAnsys 😄

File: _doc-build-linux/action.yml [hunk 7]
```diff
@@ -185,56 +185,112 @@ runs:
...
+      with:
+        level: "INFO"
+        message: >
+          Determine context.
+
+    - name: "Determine GitHub environment variables"
       shell: bash
       run: |
-        echo "EXISTS_DOC_REQUIREMENTS=$(if [ -f ${{ inputs.requirements-file }} ]; then echo 'true'; else echo 'false'; fi)" >> $GITHUB_ENV
+        if [[ -f "pyproject.toml" ]] && grep -q 'build-backend = "poetry\.core\.masonry\.api"' "pyproject.toml"; then
```
Comment (RobPasMue) [inline _doc-build-linux/action.yml:199]: Glad you also took into consideration setup.py projects (or even no-package doc projects using raw reqs files).

File: _doc-build-linux/action.yml [hunk 8]
```diff
@@ -250,6 +306,8 @@ runs:
       if: inputs.requires-xvfb == 'true'
       shell: bash
       run: |
+        source .venv/bin/activate
+        cd doc
         xvfb-run ${{ env.SPHINX_BUILD_MAKE }} -C doc html SPHINXOPTS="${{ inputs.sphinxopts }}"
         xvfb-run ${{ env.SPHINX_BUILD_MAKE }} -C doc pdf
```
Comment (RobPasMue) [inline _doc-build-linux/action.yml:312]:
```suggestion
        xvfb-run ${{ env.SPHINX_BUILD_MAKE }} -C doc html SPHINXOPTS="${{ inputs.sphinxopts }}"
        xvfb-run ${{ env.SPHINX_BUILD_MAKE }} -C doc pdf
```
If you are navigating to the "doc" folder you don't need ``-C doc``. Or viceversa. I took the second approach (not navigating) but up to you :)

**Review by RobPasMue — APPROVED** [review]
LGTM! Some minor comments

File: _doc-build-linux/action.yml [hunk 9]
```diff
@@ -185,56 +185,112 @@ runs:
...
       run: |
+        source .venv/bin/activate
         python -m pip install -r ${{ inputs.requirements-file }}
 
+    - name: "Install poetry (if required)"
+      shell: bash
+      if: env.BUILD_BACKEND == 'poetry'
+      run: |
+        source .venv/bin/activate
+        python -m pip install poetry
```
Comment (Andy-Grigg) [inline _doc-build-linux/action.yml:257]: Installing poetry into the same virtual environment that the project is installed into isn't recommended. If poetry and the running package have shared dependencies with different versions, poetry tries to upgrade/downgrade its dependencies while it's running, which often fails, especially on Windows.

The workaround we use is to use `pipx` to install poetry. Something like:

```yaml
      - name: "Add pipx/bin dir to path"
        run: echo "${{ runner.temp }}/pipx/bin" >> $GITHUB_PATH

      - name: "Install poetry and create empty virtual environment"
        run: |
          python -m pip install pipx --user
          python -m pipx install poetry
        env:
          PIPX_BIN_DIR: ${{ runner.temp }}/pipx/bin
          PIPX_HOME : ${{ runner.temp }}/pipx/home
```

Installing pipx into the `${{ runner.temp }}` directory means that the pipx installation will automatically be cleaned up between jobs on a VM. Adding the bin directory to the path first means pipx shouldn't complain that the poetry executable isn't on path.

This way, poetry is kept separate to the project, and they are free to use different versions of the same shared dependencies.

File: _doc-build-windows/action.yml [hunk 10]
```diff
@@ -212,47 +212,95 @@ runs:
...
       if: ${{ env.EXISTS_DOC_REQUIREMENTS == 'true' }}
       run: |
+        .venv\Scripts\Activate.ps1
         python -m pip install -r ${{ inputs.requirements-file }}
 
+    - name: "Install poetry (if required)"
+      shell: powershell
+      if: env.BUILD_BACKEND == 'poetry'
+      run: |
+        .venv\Scripts\Activate.ps1
```
Comment (Andy-Grigg) [inline _doc-build-windows/action.yml:281]: Same here. I won't call out every location poetry is installed, but it should always be installed in a separate environment. pipx and runner.temp is a convenient way of doing that, especially on persistent infrastructure.

**Review by Andy-Grigg — CHANGES_REQUESTED** [review]
Poetry should be installed in a separate virtual environment to the project that's being tested/built, to avoid situations in which both poetry and the project have dependencies on different versions of the same package. I've included an approach that uses pipx to get around this.

@ludovicsteinbach I know you originally raised this topic about poetry support across all actions. Can you take a look as well please?

File: _doc-build-linux/action.yml (hunk 9 above)
Comment (Andy-Grigg) [reply _doc-build-linux/action.yml:257]: This _does_ mean that pipx is installed outside of any virtual environment, but I think that's only an issue if you really care about using a specific version of pipx between different jobs, which I imagine is a pretty rare requirement.

**Review by Andy-Grigg — APPROVED** [review]
The changes look good to me. Thanks!

**Comment by SMoraisAnsys** [comment]
> The changes look good to me. Thanks!

I need to double check with some testing though :) Thanks again for the feedback !

**Comment by SMoraisAnsys** [comment]
Test closing this PR due to workflow problem

File: code-style/action.yml [hunk 11]
```diff
@@ -130,12 +195,14 @@ runs:
...
       shell: bash
       run: |
+        source .venv/bin/activate
         python -m pip install pre-commit
         pre-commit install
 
     - name: "Run pre-commit"
       shell: bash
       run: |
+        source .venv/bin/activate
```
Comment (jorgepiloto) [inline code-style/action.yml:205]: Do we need to activate the venv every time? Just curious if this is a consequence when using Poetry.

**Review by jorgepiloto — APPROVED** [review]
Thanks for the big effort here, @SMoraisAnsys. Approving. Just left one comment.

File: code-style/action.yml (hunk 11 above)
Comment (SMoraisAnsys) [reply code-style/action.yml:205]: We have to do it because we don't want to pollute / use the system installation. Otherwise, we might end up installing in the wrong place which leads to the issues in #559.
Since we work with a virtual environment, you have to activate it at each step you require it because each step is like a fresh shell. This means that not doing it in a step `B` leads to working with the system instead of your virtual environment, even if you activated it in the previous step `A`.

Comment (Andy-Grigg) [reply code-style/action.yml:205]: Alternatively, you can run the python binary directly in `./venv/bin/python` or `./venv/scripts/python` in Windows. I find this a bit awkward though, and you can end up with very long lines if you're calling pytest or sphinx with a whole load of arguments, for example.
